gdk/commands/component/BuildCommand.py

```diff
@@ -0,0 +1,425 @@
+        if build_system == "gradle":
+            return self.get_build_folders(build_folder, "build.gradle")
+        elif build_system == "maven":
+            return self.get_build_folders(build_folder, "pom.xml")
+        return {Path(utils.current_directory).joinpath(*build_folder).resolve()}
+
+    def get_build_folders(self, build_folder, build_file):
+        """
+        Recursively identifies build folders in a project.
+
+        This function makes use of build configuration files (such as pom.xml and build.gradle) and build folder
+        directories (such as target, build/libs) to identify the module directory.
+
+        Once the module directory is found, its build folder is added to the return list.
+
+        Parameters
+        ----------
+            build_folder(string): Build folder of a build system(target, build/libs)
+            build_file(string): Build configuration file of a build system (pom.xml, build.gradle)
+
+        Returns
+        -------
+            paths(set): Set of build folder paths in a multi-module project.
+        """
+        # Filter module directories which contain pom.xml or build.gradle build files.
+        set_dirs_with_build_file = set(f.parent for f in Path(utils.current_directory).rglob(build_file))
+        set_of_module_dirs = set()
+        for module_dir in set_dirs_with_build_file:
+            module_build_folder = Path(module_dir).joinpath(*build_folder).resolve()
+            # Filter module directories that also contain build folders - target/, build/libs/
+            if module_build_folder.exists():
+                set_of_module_dirs.add(module_build_folder)
+        return set_of_module_dirs
+
+    def find_artifacts_and_update_uri(self):
+        """
+        The artifact URIs in the recipe are used to identify the artifacts in local build folders of the component or on s3.
+
+        If the artifact is not found in the local build folders specific to the build system of the component, it is
+        searched on S3 with exact URI in the recipe.
+
+        Build command fails when the artifacts are neither not found in local both folders not on s3.
+
+        Parameters
+        ----------
+            None
+
+        Returns
+        -------
+            None
+        """
+        # TODO: Case insenstive recipe keys?
+        logging.info("Copying over the build artifacts to the greengrass component artifacts build folder.")
+        logging.info("Updating artifact URIs in the recipe.")
+        parsed_component_recipe = self.project_config["parsed_component_recipe"]
+        if "Manifests" not in parsed_component_recipe:
+            logging.debug("No 'Manifests' key in the recipe.")
+            return
+        build_folders = self._get_build_folder_by_build_system()
+        manifests = parsed_component_recipe["Manifests"]
+        s3_client = None
+        for manifest in manifests:
+            if "Artifacts" not in manifest:
```

Comment:
maybe make these consts as well to have a defined schema somewhere. Or use a dedicated parser

Reply:
Having a dedicated schema parser here hard as we have to keep up with the nucleus and the cloud way of validating a recipe. Though it is unlikely to change the recipe in a backward incompatible way, but in the case we do, we need to support that through gdk as well. 

So we rely on the recipe validation using the create_component_version api directly. 

Using consts is one way to go.

Reply:
Well it might also be easier to stay up to date with the Nucleus recipe schema if we have a schema to match, but I see your point